docs/unit-testing.qmd

```diff
@@ -0,0 +1,120 @@
+---
+title: Unit testing
+editor:
+  markdown:
+    wrap: sentence
+---
+
+Testing Shiny apps is important to ensure the shiny app functions as expected and to catch any bugs or errors before deployment. It helps maintain code quality, user experience, and prevents potential issues from reaching the end-users.
+
+### pytest
+
+For the example below, we will use ***pytest*** as the test framework for running our unit tests. pytest is a popular, open-source testing framework for Python. It is designed to simplify the process of writing, organizing, and running tests for Python applications and libraries.
+
+More information about ***pytest*** can be found [here](https://docs.pytest.org/en/8.2.x/contents.html).
+
+::: {.callout-tip collapse="true"}
+## Auto discovery for tests
+
+pytest automatically discovers and runs tests in your project based on a naming convention (files or functions starting with `test_*.py` or ending with `*_test.py`), eliminating the need for manual test case registration. More information about test discovery can be found [here](https://docs.pytest.org/en/8.2.x/explanation/goodpractices.html#test-discovery)
+:::
+
+
+Given a shiny app that has the following code that doubles the number for any input that a user provides.
+
+```python
+# app.py
+from shiny import render, ui
+from shiny.express import input
+
+ui.panel_title("Double your amount")
+ui.input_text("txt_box", "Enter number to double it below")
+
+
+@render.text
+def txt():
+    if input.txt_box() == "":
+        return "Please enter a number"
+    # check if input is an int or float
+    try:
+        int(input.txt_box())
+    except ValueError:
+        return "Please enter a valid number"
+    return f"n*2 is {int(value) * 2}"
```

Comment:
From https://github.com/posit-dev/py-shiny-site/pull/167#discussion_r1680134572 . Re-commented as the original file name changed

----------------

I think it'd help to have a more realistic example here. Something like filtering a dataset by category, like this?

```python
from palmerpenguins import load_penguins
from shiny.express import input, render

penguins = load_penguins()

ui.input_select("species", "Enter a species", ["Adelie", ...])

@reactive.calc
def filter_dat():
    return filter_dat_impl(penguins, input.species())

@render.data_frame
def display_dat():
    return filter_dat()


def filter_dat_impl(d, species):
    return d[d.species.isin(species)]
```